### PR TITLE
Add log streaming for spring cloud

### PR DESCRIPTION
**Pull request by Flanker32**
- Add log streaming in explorer spring cloud app node
- Add Log streaming console and correspond manager

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/StreamingLogsToolWindowManager.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import com.intellij.ui.content.ContentManagerEvent;
+import org.apache.commons.collections4.BidiMap;
+import org.apache.commons.collections4.bidimap.DualHashBidiMap;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class StreamingLogsToolWindowManager {
+
+    public static final String LOG_TOOL_WINDOW = "Azure Streaming Log";
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/StreamingLogsToolWindowManager.java:42]: private

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/StreamingLogsToolWindowManager.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        int i = 1;
+        while (resourceIdToNameMap.containsValue(result)) {
+            result = String.format("%s(%s)", resourceName, i++);
+        }
+        resourceIdToNameMap.put(resourceId, result);
+        return result;
+    }
+
+    private ToolWindow getToolWindow(Project project) {
+        if (toolWindowMap.containsKey(project)) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/StreamingLogsToolWindowManager.java:86]: where is the put method for toolWindowMap?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudStreamingLogConsoleView.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    public void startLog(InputStream inputStream) {
+        enable.set(true);
+        this.logInputStream = inputStream;
+
+        this.print("Streaming Log Start.\n", ConsoleViewContentType.SYSTEM_OUTPUT);
+        executorService = Executors.newSingleThreadExecutor();
+        executorService.submit(() -> {
+            final BufferedReader br = new BufferedReader(new InputStreamReader(logInputStream, StandardCharsets.UTF_8));
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudStreamingLogConsoleView.java:67]: use try with resource in java

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudStreamingLogConsoleView.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                    try {
+                        logInputStream.close();
+                    } catch (IOException e) {
+                        // swallow io exception when close
+                    }
+                }
+                if (executorService != null) {
+                    ThreadPoolUtils.stop(executorService, 100, TimeUnit.MICROSECONDS);
+                }
+                this.print("Streaming Log Stop.\n", ConsoleViewContentType.SYSTEM_OUTPUT);
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudStreamingLogConsoleView.java:100]: stops

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudStreamingLogManager.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+
+public class SpringCloudStreamingLogManager {
+    public static SpringCloudStreamingLogManager getInstance() {
+        return SpringCloudStreamingLogManager.SingletonHolder.INSTANCE;
+    }
+
+    public Map<String, SpringCloudStreamingLogConsoleView> consoleViewMap = new HashMap<>();
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudStreamingLogManager.java:45]: private consoleViewMap?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/StreamingLogsToolWindowManager.java (hunk 2 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/StreamingLogsToolWindowManager.java:86]: Added

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudStreamingLogConsoleView.java [hunk 6]
```diff
@@ -64,21 +64,16 @@ public void startLog(InputStream inputStream) {
...
+                    SpringCloudStreamingLogConsoleView.this.print(log + "\n", ConsoleViewContentType.NORMAL_OUTPUT);
                     Thread.sleep(50);
-                } catch (IOException e) {
-                    // swallow io exception
-                } catch (InterruptedException e) {
-                    break;
                 }
+            } catch (IOException | InterruptedException e) {
+                // swallow Exception
+                e.printStackTrace();
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudStreamingLogConsoleView.java:76]: If swallow Exception, should not print stack trace

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/serviceexplorer/azure/springcloud/SpringCloudStreamingLogsAction.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+@Name("Streaming Logs")
+public class SpringCloudStreamingLogsAction extends NodeActionListener {
+
+    private Node node;
+    private Project project;
+    private String subscriptionId;
+    private String appId;
+    private String appName;
+
+
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/serviceexplorer/azure/springcloud/SpringCloudStreamingLogsAction.java:54]: two blank lines

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/serviceexplorer/azure/springcloud/SpringCloudStreamingLogsAction.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.List;
+
+import static com.microsoft.azuretools.telemetry.TelemetryConstants.SPRING_CLOUD;
+
+@Name("Streaming Logs")
+public class SpringCloudStreamingLogsAction extends NodeActionListener {
+
+    private Node node;
+    private Project project;
+    private String subscriptionId;
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/serviceexplorer/azure/springcloud/SpringCloudStreamingLogsAction.java:50]: useless subscriptionId, appName

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudStreamingLogConsoleView.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        executorService.submit(() -> {
+            try (final BufferedReader br = new BufferedReader(new InputStreamReader(logInputStream,
+                                                                                    StandardCharsets.UTF_8));) {
+                while (enable.get()) {
+                    final String log = br.readLine();
+                    SpringCloudStreamingLogConsoleView.this.print(log + "\n", ConsoleViewContentType.NORMAL_OUTPUT);
+                    Thread.sleep(50);
+                }
+            } catch (IOException e) {
+                this.print("Streaming Log stops.\n", ConsoleViewContentType.SYSTEM_OUTPUT);
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudStreamingLogConsoleView.java:75]: Streaming Log is interrupt due to error : e.getMessage()

**Pull request merged**